### PR TITLE
Fix num static partitions to count unique partitions

### PR DESCRIPTION
**Pull request by clairelin135**
User reported an issue where partition counts were incorrect: https://elementl-workspace.slack.com/archives/C047XLJ0BRT/p1678716536609149

This was caused by duplicate static partition keys. In the next major release, we will raise an error for this. In the meantime, we'll change our partition counts logic to find the unique static partition keys.

### CODE REVIEW

**Comment by clairelin135** [comment]
This technically is a breaking change, but I think it can be classified more as a bugfix since this behavior is unsupported

**Review by sryza — COMMENTED** [review]
Is there some situation we need to worry about where old user code sends bad StaticPartitionsDefinitions over the wire to a host process that has this check and disaster ensues?

**Comment by clairelin135** [comment]
@sryza one possible situation that might happen is some external static partitions definition object contains duplicate keys and then graphQL requests fail when attempting to build a static partitions definition. The likely consequence is that the user won't be able to load the asset graph without fixing the issue and upgrading their user code.

I remember we still let the change to make escaped characters invalid go out in a dot release. I'm happy to wait until the next major release if we think it's very likely for cloud users to end up in this situation.

**Comment by johannkm** [comment]
It’d be a little tricky but I think we could test in advance if this would
affect cloud users and give a heads up. Probably will come in handy to
write a script for downloading the caches anyway. Happy to help

**Comment by clairelin135** [comment]
On second thought, I think it's probably a non-trivial number of users who might have duplicate static partition keys, given that someone reported this error. It's probably best to wait until the next major release to enforce uniqueness. For now, this PR has been amended to only count the number of unique partitions instead to display the correct information in Dagit.

File: python_modules/dagster/dagster/_core/definitions/partition.py [hunk 1]
```diff
@@ -423,6 +423,13 @@ def __eq__(self, other) -> bool:
     def __repr__(self) -> str:
         return f"{type(self).__name__}(partition_keys={[p.name for p in self._partitions]})"
 
+    def get_num_partitions(
+        self,
+        current_time: Optional[datetime] = None,
+        dynamic_partitions_store: Optional[DynamicPartitionsStore] = None,
+    ) -> int:
+        return len(set(self.get_partition_keys(current_time, dynamic_partitions_store)))
```
Comment (johannkm) [inline python_modules/dagster/dagster/_core/definitions/partition.py:431]: could add a comment about why we need a set

**Review by johannkm — APPROVED** [review]
Does this leave the default implementation unused? could remove

**Comment by clairelin135** [comment]
The default implementation is still used for multipartitions

**Comment by clairelin135** [comment]
Hm.. the counts could also be off for the multipartitioned case. I'll add a similar function for the multipartitioned case to return unique counts, but I still think we should retain the original default function in case if anybody has a custom partitions definition.

**Pull request merged**